cpp/src/arrow/util/thread_pool_test.cc

```diff
@@ -189,11 +189,7 @@ TEST_P(TestRunSynchronously, SpawnMoreNested) {
   auto top_level_task = [&](Executor* executor) -> Future<> {
     auto fut_a = DeferNotOk(executor->Submit([&] { nested_ran++; }));
     auto fut_b = DeferNotOk(executor->Submit([&] { nested_ran++; }));
-    return AllComplete({fut_a, fut_b})
-        .Then([&](const Result<arrow::detail::Empty>& result) {
-          nested_ran++;
-          return result;
-        });
+    return AllComplete({fut_a, fut_b}).Then([&]() { nested_ran++; });
```

Comment:
I'm curious why the old version took a `Result` and returned it. Was there a discrepancy between `Future<T>::Then` and `Future<Empty>::Then`?

Reply:
What a continuation returns is a separate decision from what it accepts.  However, returning a result here was essentially a no-op before and would still be one now.

The first argument (or only argument in the 1-argument overload) to `Then` will only be called on success and it is given `result.ValueOrDie()` so accepting a `Result` is somewhat counter-productive.  This is different from `AddCallback` which is called on success or failure.

It would be good to make it a compile error to accept a `Result<T>` in all cases of `Then` to avoid this confusion.  The current state of the PR will make a compile error for `Future<Empty>::Then` (which is why this was caught) so I'll update it to catch the `Future<T>::Then` case as well.

Reply:
I took a stab at preventing `Then` callbacks accepting `Result<T>` but didn't make much headway and so I'd prefer to defer that to a future PR.  The problem is that we have `Args` in `ContinueFuture` but no `T` to check if `Args[0] == Result<T>`.  In the `Then` method we have `T` but no idea what `Args` are.

I suspect there is someway to test if `Args[0]` is a `Result<?>` but I tried a few things and didn't figure it out.

Reply:
Thanks, this is great.  `call_traits` doesn't work for function pointers so that caused one call to be slightly less readable (there was already a TODO but I promoted it to a full JIRA issue ARROW-12655.

However, I think it's worth it.  This caught several places we were accepting `Result` and one in `async_generator.h` that was potentially a bug.